<a href="https://colab.research.google.com/github/mlacasa/1BATXILLERAT/blob/main/Pr%C3%A0cticaBasedeDades.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PRÀCTICA ASSIGNATURA MATEMÀTIQUES APLICADA 1 BAT
## Marcos Lacasa

1- Carreguem l'arxiu "Distrubución Salarios" de open-acces de l'agència tributària.



In [1]:
# Importem llibreria pandas
import pandas as pd

In [2]:
# definim el path des del nostre drive i carreguem el format csv.
path = '/content/drive/MyDrive/Colab Notebooks/BAT/data2.csv'
data = pd.read_csv(path)

In [3]:
data.head()

ID         2001         2002         2003         2004         2005  \
0   1   29,541,615   27,645,802   29,437,929   27,855,280   31,801,465   
1   2   72,965,628   81,467,945   87,340,372   68,217,591   71,854,670   
2   3  118,397,894  117,964,368  113,769,342  102,460,947  109,068,935   
3   4  149,288,420  145,505,890  153,889,058  144,327,889  154,811,651   
4   5  187,829,221  179,214,060  177,103,986  164,526,172  179,556,090   

          2006         2007         2008         2009  ... 2013.4 2014.4  \
0   31,584,347   25,499,488   27,306,225   28,879,391  ...     93     94   
1   72,813,778   59,420,277   62,227,592   67,221,958  ...    296    295   
2  108,110,010   88,723,845   89,901,678   99,150,785  ...    498    497   
3  154,349,596  130,103,812  129,261,037  140,721,048  ...    702    701   
4  174,609,190  143,789,890  140,636,251  159,501,515  ...    896    898   

  2015.4 2016.4 2017.4 2018.4 2019.4 2020.4 2021.4 2022.4  
0     95     96     97     98     98    100     99     98  
1    296    296    296    296    295    296    297    292  
2    498    500    498    499    498    499    498    497  
3    700    699    697    697    700    700    699    699  
4    897    895    897    896    898    900    899    899  

[5 rows x 111 columns]

In [4]:
data.shape

(400, 111)

### OBSERVACIONS

1- Comprobem que les dades son 'objectes' i els nombres tenen 2 commes.
2- Hem de suprimir les commes de totes les columnes.

Hi ha 400 files (una per cada tram) i 111 columnes (que son 5 categoríes per cada 22 anys analitzats.

In [6]:
# Convertim totes les columnes en "string" per poder utilitzar el mètode "replace"
data = data.astype(str)

In [7]:
# Treiem les commes però ho fem amb un loop utilitzant 'iloc' de pandas
for i in range(1,111):
  data.iloc[:,i]=data.iloc[:,i].str.replace(',','')

### documentació iloc de pandas

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html

Visualitzar!!

In [8]:
# Seleccionem només les 3 primeres categoríes
# Les altres dues, son vectors LD, 22x3+1 = 67 colummes
data1 = data.iloc[:,0:67]

In [10]:
# Convertim totes les columnes com nombres naturals (enters)
data1 = data1.astype(int)

In [11]:
# comprobem i visualitzem les 5 primers registres
data1.head()

ID       2001       2002       2003       2004       2005       2006  \
0   1   29541615   27645802   29437929   27855280   31801465   31584347   
1   2   72965628   81467945   87340372   68217591   71854670   72813778   
2   3  118397894  117964368  113769342  102460947  109068935  108110010   
3   4  149288420  145505890  153889058  144327889  154811651  154349596   
4   5  187829221  179214060  177103986  164526172  179556090  174609190   

        2007       2008       2009  ...  2013.2  2014.2  2015.2  2016.2  \
0   25499488   27306225   28879391  ...  384262  369837  360424  359341   
1   59420277   62227592   67221958  ...  267525  263879  269745  270013   
2   88723845   89901678   99150785  ...  227742  230155  235644  242355   
3  130103812  129261037  140721048  ...  219984  217609  228211  242164   
4  143789890  140636251  159501515  ...  204015  202091  214043  214446   

   2017.2  2018.2  2019.2  2020.2  2021.2  2022.2  
0  336325  328778  310946  270572  276001  244376  
1  248893  233521  210956  194006  200752  186871  
2  226109  210252  180277  170422  164307  155482  
3  208225  192375  172495  164014  157221  149360  
4  191359  186405  164155  164394  146657  139655  

[5 rows x 67 columns]

In [12]:
data1.dtypes # Avaluem el tipus de cada columna

ID        int64
2001      int64
2002      int64
2003      int64
2004      int64
          ...  
2018.2    int64
2019.2    int64
2020.2    int64
2021.2    int64
2022.2    int64
Length: 67, dtype: object

## Preparem el format per analitzar i visualitzar

Hem de crear la dataframe amb el disseny que ens interessa

In [22]:
# Vector amb els anys, del 2001-2022
years = list(range(2001,2023))
print("el primer any és: ",years[0]," i l'últim és ", years[21])

el primer any és:  2001  i l'últim és  2022


In [24]:
# Els noms de les columnes de la nostre DataFrame
nomsCol = ['rank', 'value', 'year', 'category']

In [25]:
data2 = pd.DataFrame(columns = nomsCol)

In [26]:
data2

Empty DataFrame
Columns: [rank, value, year, category]
Index: []

In [ ]:
# Les 3 categories son:
cat = ['Salary', 'tax', 'employees']

## Exercici

1- Configurar la 'data2' de la següent manera:

- Ha de tenir tots els valors de 'data1' disposats en 4 columnes.
- Heu de dissenyar una funció per tal de recòrrer les columnes i els valors.
- cada valor ha d'estar definit per un rank, value, year i category.
- Les 3 categoríes son al vector 'cat'.
- Els anys són al vector 'years'.

Exemple del 1er registre:

rank | value | year | category  
--- | --- |--- |--- |
1|29541615|2001|Salary


### Notes

- Les columnes amb els anys pertànyen:\
1- 2001 -> Salary\
2- 2001.1 -> tax\
3- 2001.2 -> employees

### Indicacions

Realitzar els loops amb diferents "for" per recòrrer les columnes. Utilitzar adequadament el mètode 'iloc' de pandas.

Recordar:


```
df.iloc[1:3, 0:3]
```
vol dir, escull les files 1,2 i les columnes 0,1,2.


```
df.iloc[:, 1]
```
vol dir, escull totes les files de la columna 2.


In [ ]:
# Indicar la funció
